File: solver/cookies.test.js
// Run with: npm test --prefix solver   (or: node --test solver/*.test.js)
//
// Not `node --test solver/`: an explicit directory makes the runner execute
// every .js under it, index.js included, which is a CLI and not a test.
//
// Covers the fallback matcher in cookies.js. The primary path (page.cookies)
// is the browser's own matching and needs a browser to exercise; this is the
// code that runs when that method is unavailable, and a mistake here sends a
// cookie to a host it was never issued to.

import test from "node:test";
import assert from "node:assert/strict";
import { cookieInScope, cookiesForUrl, targetScope } from "./cookies.js";

const scope = (url) => targetScope(url);

test("host-only cookies match their exact host", () => {
  const s = scope("https://example.com/");
  assert.ok(cookieInScope({ domain: "example.com", path: "/" }, s));
  assert.ok(!cookieInScope({ domain: "other.com", path: "/" }, s));
});

test("domain cookies cover subdomains, in either spelling", () => {
  const s = scope("https://www.example.com/");
  assert.ok(cookieInScope({ domain: ".example.com", path: "/" }, s));
  assert.ok(cookieInScope({ domain: "example.com", path: "/" }, s));
  assert.ok(cookieInScope({ domain: "www.example.com", path: "/" }, s));
});

test("a subdomain's cookie does not leak to the parent", () => {
  const s = scope("https://example.com/");
  assert.ok(!cookieInScope({ domain: "www.example.com", path: "/" }, s));
});

test("suffix matching stops at the dot boundary", () => {
  // The bug this guards: endsWith("example.com") alone would hand
  // example.com's cf_clearance to notexample.com and vice versa.
  const s = scope("https://notexample.com/");
  assert.ok(!cookieInScope({ domain: "example.com", path: "/" }, s));
  assert.ok(!cookieInScope({ domain: "notexample.com", path: "/" }, scope("https://example.com/")));
});

test("the challenge iframe's origin is out of scope", () => {
  const s = scope("https://shop.example.com/checkout");
  assert.ok(!cookieInScope({ domain: "challenges.cloudflare.com", path: "/" }, s));
});

test("domain matching ignores case", () => {
  const s = scope("https://WWW.Example.COM/");
  assert.ok(cookieInScope({ domain: "Example.com", path: "/" }, s));
});

test("path matching respects segment boundaries", () => {
  const s = scope("https://example.com/admin/users");
  assert.ok(cookieInScope({ domain: "example.com", path: "/" }, s));
  assert.ok(cookieInScope({ domain: "example.com", path: "/admin" }, s));
  assert.ok(cookieInScope({ domain: "example.com", path: "/admin/" }, s));
  assert.ok(cookieInScope({ domain: "example.com", path: "/admin/users" }, s));
  assert.ok(!cookieInScope({ domain: "example.com", path: "/administrator" }, s));
  assert.ok(!cookieInScope({ domain: "example.com", path: "/other" }, s));
});

test("a cookie with no domain is never in scope", () => {
  assert.ok(!cookieInScope({ path: "/" }, scope("https://example.com/")));
  assert.ok(!cookieInScope({ domain: "", path: "/" }, scope("https://example.com/")));
});

test("cookiesForUrl reads the profile jar and filters it", async () => {
  // The jar carries a cf_clearance for another origin, and the unfiltered list
  // would have handed that back as the solve.
  const browser = {
    cookies: async () => [
      { name: "cf_clearance", value: "third-party", domain: "challenges.cloudflare.com", path: "/" },
      { name: "junk", value: "1", domain: "cdn.other.net", path: "/" },
      { name: "cf_clearance", value: "target", domain: "example.com", path: "/" },
    ],
  };
  const got = await cookiesForUrl(browser, {}, "https://example.com/");
  assert.equal(got.length, 1);
  assert.equal(got[0].value, "target");
});

// The regression this pins: page.cookies was the primary read, and it returns
// an empty list — not an error — when the page handle is no longer attached to
// the target that did the solving, which is what puppeteer-real-browser's
// targetcreated rewrapping produces. A live run reported "cookie_list":[] for a
// site that had just issued a cf_clearance.
test("an empty page read does not mask the profile jar", async () => {
  const page = { cookies: async () => [] };
  const browser = {
    cookies: async () => [
      { name: "cf_clearance", value: "target", domain: "example.com", path: "/" },
      { name: "__cf_bm", value: "bm", domain: "example.com", path: "/" },
    ],
  };
  const got = await cookiesForUrl(browser, page, "https://example.com/");
  assert.equal(got.length, 2, "the profile jar was not consulted");
  assert.ok(got.some((c) => c.name === "cf_clearance"));
});

test("cookiesForUrl falls back to the page when the browser has no jar API", async () => {
  // puppeteer below 23.7 has no Browser.cookies at all.
  const seen = [];
  const page = {
    cookies: async (...urls) => {
      seen.push(urls);
      return [{ name: "sess", value: "1", domain: "example.com", path: "/" }];
    },
  };
  const got = await cookiesForUrl({}, page, "https://example.com/");
  assert.deepEqual(seen, [["https://example.com/"]], "the target url was not passed through");
  assert.equal(got[0].name, "sess");
});

test("cookiesForUrl falls back to the page when the jar read throws", async () => {
  const browser = {
    cookies: async () => {
      throw new Error("protocol error");
    },
  };
  const page = {
    cookies: async () => [{ name: "sess", value: "1", domain: "example.com", path: "/" }],
  };
  const got = await cookiesForUrl(browser, page, "https://example.com/");
  assert.equal(got[0].name, "sess");
});

test("cookiesForUrl yields nothing rather than throwing on a bad url", async () => {
  const browser = { cookies: async () => [{ name: "a", value: "1", domain: "example.com" }] };
  assert.deepEqual(await cookiesForUrl(browser, {}, "not a url"), []);
});

File: internal/solver/cookies_test.go
package solver

import (
	"testing"

	"github.com/JSInvasor/Gohttp-clientfingerprintemulateandfastest/internal/cdp"
)

func TestTargetScope(t *testing.T) {
	cases := []struct {
		url  string
		host string
		path string
		ok   bool
	}{
		{"https://Example.COM/a/b", "example.com", "/a/b", true},
		{"https://example.com", "example.com", "/", true},
		{"http://sub.example.com:8443/x", "sub.example.com", "/x", true},
		{"not a url", "", "", false},
		{"", "", "", false},
	}
	for _, tc := range cases {
		scope, ok := TargetScope(tc.url)
		if ok != tc.ok {
			t.Errorf("TargetScope(%q) ok = %v, want %v", tc.url, ok, tc.ok)
			continue
		}
		if ok && (scope.Host != tc.host || scope.Path != tc.path) {
			t.Errorf("TargetScope(%q) = %+v, want %s %s", tc.url, scope, tc.host, tc.path)
		}
	}
}

// The rules that keep one host's cf_clearance from being reported as another's.
func TestCookieInScope(t *testing.T) {
	scope, _ := TargetScope("https://shop.example.com/admin")

	cases := []struct {
		name   string
		cookie cdp.Cookie
		want   bool
	}{
		{"exact host", cdp.Cookie{Domain: "shop.example.com", Path: "/"}, true},
		{"parent domain", cdp.Cookie{Domain: "example.com", Path: "/"}, true},
		{"leading dot is the same thing", cdp.Cookie{Domain: ".example.com", Path: "/"}, true},
		{"case is not significant", cdp.Cookie{Domain: "Example.COM", Path: "/"}, true},
		// The one that matters: a bare suffix match would let this collect the
		// target's cookies.
		{"suffix that is not a dot boundary", cdp.Cookie{Domain: "evil-example.com", Path: "/"}, false},
		{"unrelated host", cdp.Cookie{Domain: "other.com", Path: "/"}, false},
		{"a subdomain is not a parent", cdp.Cookie{Domain: "deep.shop.example.com", Path: "/"}, false},
		{"empty domain", cdp.Cookie{Domain: "", Path: "/"}, false},

		{"matching path", cdp.Cookie{Domain: "shop.example.com", Path: "/admin"}, true},
		{"empty path is root", cdp.Cookie{Domain: "shop.example.com", Path: ""}, true},
		{"a longer path does not match", cdp.Cookie{Domain: "shop.example.com", Path: "/admin/deep"}, false},
		// "/admin" covers "/admin/x" but not "/administrator".
		{"prefix that is not a path boundary", cdp.Cookie{Domain: "shop.example.com", Path: "/ad"}, false},
	}
	for _, tc := range cases {
		if got := CookieInScope(tc.cookie, scope); got != tc.want {
			t.Errorf("%s: CookieInScope(%+v) = %v, want %v", tc.name, tc.cookie, got, tc.want)
		}
	}

	deeper, _ := TargetScope("https://shop.example.com/admin/users")
	if !CookieInScope(cdp.Cookie{Domain: "shop.example.com", Path: "/admin"}, deeper) {
		t.Error(`"/admin" should cover "/admin/users"`)
	}
	if CookieInScope(cdp.Cookie{Domain: "shop.example.com", Path: "/administrator"}, deeper) {
		t.Error(`"/administrator" must not cover "/admin/users"`)
	}
}

// The failure this whole file exists to prevent, end to end: a challenge run
// navigates cross-origin and back, the challenge host sets its own cf_clearance
// on the way through, and the unfiltered jar reports the wrong one as the solve.
func TestCookiesForURLDropsAnotherHostsClearance(t *testing.T) {
	jar := []cdp.Cookie{
		{Name: "cf_clearance", Value: "THIRD_PARTY", Domain: "challenges.cloudflare.com", Path: "/"},
		{Name: "tp_junk", Value: "1", Domain: "tracker.example.net", Path: "/"},
		{Name: "cf_clearance", Value: "REAL", Domain: "target.example", Path: "/"},
		{Name: "__cf_bm", Value: "bm1", Domain: ".target.example", Path: "/"},
	}

	got := CookiesForURL(jar, "https://target.example/")
	if len(got) != 2 {
		t.Fatalf("kept %d cookies, want 2: %+v", len(got), got)
	}
	for _, c := range got {
		if c.Name == "cf_clearance" && c.Value != "REAL" {
			t.Fatalf("reported %q as the clearance; the third party's cookie survived the filter", c.Value)
		}
	}

	header := CookieHeader(got)
	if header != "cf_clearance=REAL; __cf_bm=bm1" {
		t.Errorf("CookieHeader = %q", header)
	}
	if !HasClearance(got) {
		t.Error("HasClearance did not see the target's own clearance")
	}
	if HasClearance(CookiesForURL(jar, "https://unrelated.example/")) {
		t.Error("a host with no cookies of its own was reported as cleared")
	}
}

func TestCookiesForURLRejectsAnUnparseableTarget(t *testing.T) {
	jar := []cdp.Cookie{{Name: "a", Domain: "example.com", Path: "/"}}
	if got := CookiesForURL(jar, "not a url"); len(got) != 0 {
		t.Errorf("CookiesForURL(garbage) = %+v, want nothing", got)
	}
}
